calico_containers/pycalico/ipam.py

```diff
@@ -455,10 +459,18 @@ def _auto_assign(self, ip_version, num, handle_id,
                 _log.info("Ran out of affine blocks for %s in pool %s",
                           hostname, pool)
                 break
-            ips = self._auto_assign_block(block_id,
-                                          num_remaining,
-                                          handle_id,
-                                          attributes)
+            try:
+                ips = self._auto_assign_block(block_id,
+                                              num_remaining,
+                                              handle_id,
+                                              attributes)
+            except (KeyError, NoHostAffinityWarning):
+                # In certain rare race conditions, _get_affine_blocks above
+                # can return block_ids that don't exist or don't actually have
+                # affinity to this host (due to multiple IPAM clients on this
+                # host running simultaneously).  If that happens, just move to
+                # the next one.
```

Comment:
In the event of a KeyError don't we want to retry from the beginning again?  i.e. get the affine blocks again and then retry (but perhaps skipping the blocks we already tried)?  Otherwise we potentially assign another affine block when it isn't necessary.